### PR TITLE
Limit the max request size to be accepted

### LINKED ISSUE #785: Bad Frame (with large length) can cause grpc.Server to explode
While experimenting with a custom transport (to be touched on another type), I found that if you send a bad header  (in `parser`), you can irrecoverably kill the server.

rpc_util.go:231
```go
func (p *parser) recvMsg() (pf payloadFormat, msg []byte, err error) {
	if _, err := io.ReadFull(p.r, p.header[:]); err != nil {
		return 0, nil, err
	}

	pf = payloadFormat(p.header[0])
	length := binary.BigEndian.Uint32(p.header[1:])

	if length == 0 {
		return pf, nil, nil
	}
	grpclog.Printf("Got Header %v, length: %d", p.header, length)
	// TODO(bradfitz,zhaoq): garbage. reuse buffer after proto decoding instead
	// of making it for each message:
	msg = make([]byte, int(length))
```

If you get something like 
```grpc: 2016/07/28 14:56:34 Got Header [0 120 48 48 0], length: 2016423936```

```
goroutine 51 [running]:
runtime.systemstack_switch()
	/usr/local/go-1.7beta2/src/runtime/asm_amd64.s:252 fp=0xc4201cd9b8 sp=0xc4201cd9b0
runtime.mallocgc(0x78303000, 0x751800, 0x514501, 0xc42001b4a0)
	/usr/local/go-1.7beta2/src/runtime/malloc.go:670 +0x8f7 fp=0xc4201cda58 sp=0xc4201cd9b8
runtime.makeslice(0x751800, 0x78303000, 0x78303000, 0x2, 0x2, 0x5)
	/usr/local/go-1.7beta2/src/runtime/slice.go:57 +0x7b fp=0xc4201cdab0 sp=0xc4201cda58
google.golang.org/grpc.(*parser).recvMsg(0xc420181da0, 0xc420181dc0, 0xc4201cddc8, 0xc420159200, 0xc42002dbd1, 0x3, 0x27)
	/home/michal/Private/code/nonroot/src/google.golang.org/grpc/rpc_util.go:245 +0x1c5 fp=0xc4201cdb48 sp=0xc4201cdab0
google.golang.org/grpc.(*Server).processUnaryRPC(0xc420093720, 0x983ac0, 0xc42001cfc0, 0xc4200cea50, 0xc420158900, 0x99aa58, 0xc420159200, 0x0, 0x0)
	/home/michal/Private/code/nonroot/src/google.golang.org/grpc/server.go:530 +0x1d4 fp=0xc4201cdd90 sp=0xc4201cdb48
google.golang.org/grpc.(*Server).handleStream(0xc420093720, 0x983ac0, 0xc42001cfc0, 0xc4200cea50, 0xc420159200)
	/home/michal/Private/code/nonroot/src/google.golang.org/grpc/server.go:744 +0x6b0 fp=0xc4201cdf50 sp=0xc4201cdd90
google.golang.org/grpc.(*Server).serveStreams.func1.1(0xc420184a30, 0xc420093720, 0x983ac0, 0xc42001cfc0, 0xc4200cea50)
	/home/michal/Private/code/nonroot/src/google.golang.org/grpc/server.go:408 +0xab fp=0xc4201cdf88 sp=0xc4201cdf50
runtime.goexit()
	/usr/local/go-1.7beta2/src/runtime/asm_amd64.s:2086 +0x1 fp=0xc4201cdf90 sp=0xc4201cdf88
created by google.golang.org/grpc.(*Server).serveStreams.func1
	/home/michal/Private/code/nonroot/src/google.golang.org/grpc/server.go:409 +0xa3
```

I think gRPC servers should guard against too large allocations, putting a limit on the maximum. `grpc-java` seems to be doing `4MB`

https://github.com/grpc/grpc-java/issues/1676

### PR DESCRIPTION
**Pull request by iamqizhao**
Make it consistent with Java and C++ impl.

also fixes #785 

TODO: client side memory guard.

### CODE REVIEW

**Comment by menghanl** [comment]
LGTM.

**Comment by iamqizhao** [comment]
@tamird Yes, it should be. I will do an entire pass on this file in a separate PR. Most of tests exist before codes became stringers.

File: rpc_util.go [hunk 1]
```diff
@@ -319,11 +322,16 @@ func recv(p *parser, c Codec, s *transport.Stream, dc Decompressor, m interface{
 	if pf == compressionMade {
 		d, err = dc.Do(bytes.NewReader(d))
 		if err != nil {
-			return transport.StreamErrorf(codes.Internal, "grpc: failed to decompress the received message %v", err)
+			return Errorf(codes.Internal, "grpc: failed to decompress the received message %v", err)
 		}
 	}
+	if len(d) > maxMsgSize {
```
Comment (c4milo) [inline rpc_util.go]: shouldn't this validation be done before decompressing the message?

Comment (c4milo) [inline rpc_util.go]: additionally, it seems the message size validation is also being done in the parser: https://github.com/grpc/grpc-go/blob/master/rpc_util.go#L253

**Pull request merged**